subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java

```diff
@@ -453,8 +451,7 @@ public ClassMetadata(Class<?> type) {
          * Determines the concrete return type of the given method, resolving any type parameters.
          */
         public MethodMetadata resolveTypeVariables(Method method) {
-            Type returnType = method.getGenericReturnType();
-            Type resolvedReturnType = JavaReflectionUtil.hasTypeVariable(returnType) ? TypeToken.of(type).method(method).getReturnType().getType() : returnType;
+            Type resolvedReturnType = JavaPropertyReflectionUtil.resolveMethodReturnType(type, method);
```

Comment:
Good call! We could also call the method `resolvePropertyType()` if it's in `JavaPropertyReflectionUtil`. But I don't feel strongly either way.